packages/account/src/providers/provider.ts

```diff
@@ -773,48 +782,56 @@ export default class Provider {
       estimatePredicates = true,
       resourcesOwner,
     }: TransactionCostParams = {}
-  ): Promise<TransactionCost> {
-    const transactionRequest = transactionRequestify(clone(transactionRequestLike));
+  ): Promise<
+    TransactionCost & {
+      estimatedInputs: TransactionRequest['inputs'];
+      estimatedOutputs: TransactionRequest['outputs'];
+    }
+  > {
+    const txRequestClone = clone(transactionRequestify(transactionRequestLike));
     const chainInfo = this.getChain();
     const { gasPriceFactor, minGasPrice, maxGasPerTx } = this.getGasConfig();
-    const gasPrice = max(transactionRequest.gasPrice, minGasPrice);
-    const isScriptTransaction = transactionRequest.type === TransactionType.Script;
+    const gasPrice = max(txRequestClone.gasPrice, minGasPrice);
+    const isScriptTransaction = txRequestClone.type === TransactionType.Script;
+
+    /**
+     * Fund with fake UTXOs to avoid not enough funds error
+     */
+    // Getting coin quantities from amounts being transferred
+    const coinOutputsQuantities = txRequestClone.getCoinOutputsQuantities();
+    // Combining coin quantities from amounts being transferred and forwarding to contracts
+    const allQuantities = mergeQuantities(coinOutputsQuantities, forwardingQuantities);
+    // Funding transaction with fake utxos
+    txRequestClone.fundWithFakeUtxos(allQuantities, resourcesOwner?.address);
 
     /**
      * Estimate predicates gasUsed
      */
-    if (transactionRequest.hasPredicateInput() && estimatePredicates) {
+    if (estimatePredicates) {
       // Remove gasLimit to avoid gasLimit when estimating predicates
       if (isScriptTransaction) {
-        transactionRequest.gasLimit = bn(0);
+        txRequestClone.gasLimit = bn(0);
       }
-      await this.estimatePredicates(transactionRequest);
+
+      if (resourcesOwner && 'populateTransactionPredicateData' in resourcesOwner) {
```

Comment:
What is this check? Any risk in us not having `populateTransactionPredicateData` present?

Reply:
A quick doc may be helpful here for some of this very specific logic, prevalent in predicates.

Reply:
https://github.com/FuelLabs/fuels-ts/pull/1767/commits/5e7fd11192c46cdcd242fb0ff68eb50d016c1275